Separate sip code

annekainicUSDS opened the pull request:
Connected to https://github.com/usds/us-forms-system/issues/2.

#### Purpose
This PR refactors the save in progress (SiP) code so that it is completely separate from the core library code. Our SiP implementation is very specific to Vets.gov, so it made sense to disentangle that code from the rest of the library code and create clearer boundaries for how SiP functionality would be integrated with the core library code, whatever the implementation.

This is work was done originally in what will eventually be an abstracted version of the forms library code (https://github.com/department-of-veterans-affairs/us-forms-system/pull/6), but we decided to pull it back into vets-website as long as the code lives in two places.

#### Changes made
There are three main changes to keep track of:
1. Previously, `FormApp` was the **top level component**, and from within `FormApp` `RoutedSavableApp`, the component to handle SiP logic, was rendered. We decided to invert this relationship so that apps that need SiP functionality will instead use `RoutedSavableApp` as their top level component. `RoutedSavableApp` is purely responsible for the logic behind what page or content the user is redirected to based on the status of their prefilled application, and `FormApp` will be responsible for rendering form components like nav, title, and form content. Apps that do not need SiP functionality will continue to use `FormApp`.
2. The **reducer** code has been refactored so that there is a basic `createSchemaFormReducer` function that no longer needs to have knowledge of any SiP logic. Instead a new reducer function, `createSaveInProgressFormReducer` has been created that will overwrite the state and reducers with specific ones relevant to SiP.
3. The **routes** have been refactored so that the essential routes are created without knowledge of SiP, and then a new function, `addSaveInProgressRoutes`, adds the SiP routes where necessary.

#### Forms this Affects
I've made changes to both the library code and all apps where prefill is enabled. They are:
- [ ] Burials
- [ ] disability 526EZ
- [ ] edu 1990
- [ ] edu 1990e
- [ ] edu 1990n
- [ ] edu 1995
- [ ] edu 5490
- [ ] edu 5495
- [ ] pensions
- [ ] pre-need
- [ ] veteran-representative
- [ ] vic-v2
- [ ] vre Chapter31
- [ ] vre Chapter36

I deliberately did not make changes to disability 686 form because prefill is not enabled for that one, so it should remain the same. If anything about the above list looks incorrect, please let me know.

I need help in reviewing and testing this new implementation. I've done some preliminary testing, but I'm not familiar with many of the apps. I'd like to divide up these forms to make sure that each is reviewed and tested by someone who knows it best. Once it has been reviewed and tested thoroughly, please check off the form from the above list so we can keep track of ones that are good to go.

Let me know if there are any questions, thanks for your help!

Commit:
Refactor SiP code so that it operates completely separately from base library code

Commit:
Add changes to HCA

Commit:
Convert all forms with prefill enabled to using new method for incorporating SiP functionality

Commit:
Fix broken paths

jbalboni reviewed (commented):
Looks good so far, I will test it locally later today.

rianfowler commented on src/js/common/schemaform/save-in-progress/helpers.js:
```diff
@@ -0,0 +1,65 @@
+  const newRoutes = routes;
+
+  newRoutes.forEach((route) => {
+    const indexOfRoute = newRoutes.indexOf(route);
```
`forEach` passes the current value `index` as the second argument to this callback.

annekainicUSDS replied:
ah, good catch thanks

rianfowler commented on src/js/common/schemaform/save-in-progress/helpers.js:
```diff
@@ -0,0 +1,65 @@
+import SaveInProgressErrorPage from './SaveInProgressErrorPage';
+
+export function addSaveInProgressRoutes(config, routes) {
+  const protectedRoutes = ['introduction', 'review-and-submit', 'confirmation', '*'];
```
What are these routes protected from?

annekainicUSDS replied:
@rianfowler from being overwritten by rewriting the component to render for them (line 23). They have their own components to render, or get redirected.

annekainicUSDS commented on src/js/common/schemaform/helpers.js:
```diff
@@ -73,7 +69,7 @@ export function createPageList(formConfig, formPages) {
   const formPages = createFormPageList(formConfig);
   const pageList = createPageList(formConfig, formPages);
-  const Page = formConfig.disableSave ? FormPage : RoutedSavablePage;
+  const Page = FormPage;
```
good point, will change that

jbalboni commented:
@annekainicUSDS Testing it out locally, it looks like the form saved page and the resume page aren't working correctly. If you click "Finish application later," you're sent back to the intro page. Similarly, if you click "Continue Application" from the profile page, you get put on the intro page. This is all controlled in RoutedSavableApp, so the issue is likely in there somewhere.

Commit:
Refactor addSaveInProgressRoutes

Commit:
Move new helpers file to new directory

cvalarida reviewed (commented):
Looks fine so far. I'll be firing it up and walking through a form or two to see the issue @jbalboni discovered.

Commit:
Fix lint errors

annekainicUSDS commented on src/js/common/schemaform/save-in-progress/RoutedSavableApp.jsx:
```diff
@@ -193,27 +191,22 @@ class RoutedSavableApp extends React.Component {
       content = <LoadingIndicator message="Retrieving your profile information..."/>;
     } else if (!isInProgress(trimmedPathname)) {
-      content = children;
+      content = (
```
yeah... I'm still trying to figure out why it was different in the first place, I don't feel totally confident that I've understood this well enough yet to remove it.

annekainicUSDS commented:
I've been trying to test this locally and figure out why the `/resume` and `/form-saved` pages are redirecting to `/introduction`. What I've noticed is that before `HealthCareEntry` (within `HealthCareApp` component) is rendered, `this.props.location` is reset to `introduction` from `resume` or whatever it was before. I don't yet know what's causing this prop to be reset, and why it wasn't being reset before. But wanted to provide an update of what I've figured out so far.

Commit:
Fix routing issue for form-saved and resume routes

Commit:
Fix form-saved styling

Commit:
Fix styling issue with wrapping FormApp

annekainicUSDS commented:
Fixed the bug with the `/form-saved` and `/resume` routing. Also had to change some other styles for those pages. Now fixing the broken unit tests.

Commit:
Fix broken tests

Commit:
Add additional unit tests for new code refactor

annekainicUSDS commented:
@jbalboni this should be ready to take another look at whenever you have a chance. Fixed the issue with the routing and the broken unit tests (also added some new ones for the refactor). I can see that the Jenkins build failed, but unfortunately the link to see what in particular failed is not working for me. Trying to run the e2e tests locally also is not working, so will continue to dig into to get those running locally and see if that's the cause of the Jenkins fail.

jbalboni commented on src/applications/common/schemaform/save-in-progress/helpers.js:
```diff
@@ -0,0 +1,63 @@
+    // rewrite page component
+    if (!protectedRoutes.has(route.path)) {
+      newRoute = Object.assign({}, route, { component: RoutedSavablePage });
+      if (index !== -1) {
```
When is the index set to `-1`?

annekainicUSDS replied:
I can't remember why I added this... and looking at it it doesn't make sense. Going to just remove it.

jbalboni commented:
@annekainicUSDS If you're still not able to see the Jenkins build failure, here's what I see:

```
 ✖ src/platform/forms/tests/save-in-progress/01-sip-load-fail.e2e.spec
   - Begin application (7.316s)
   Expected element <.js-test-location> to have attribute "data-location" not contain: "/introduction" in 5000ms  - expected "not contain '/introduction'" but got: "/introduction"
       at Object.expectNavigateAwayFrom (/application/src/platform/testing/e2e/helpers.js:85:17)
       at /application/src/platform/forms/tests/save-in-progress/01-sip-load-fail.e2e.spec.js:31:16
       at Object.BeginApplication [as Begin application] (/application/src/platform/testing/e2e/helpers.js:77:7)
```

jbalboni commented:
This looks good to me, though I think we should update the 686 form, too. I can give it a final look once the tests are fixed.

Commit:
Fix failing e2e test and rename function

jbalboni commented:
686 has save in progress enabled, so it should be the same set of changes as the other forms.

annekainicUSDS commented on src/applications/common/schemaform/save-in-progress/helpers.js:
```diff
@@ -0,0 +1,63 @@
+import FormSaved from './FormSaved';
+import SaveInProgressErrorPage from './SaveInProgressErrorPage';
+
+export function addSaveInProgressRoutes(config) {
```
Changed

Commit:
Add changes to 686

Commit:
remove unnecessary check for index

Commit:
Fix broken tests

annekainicUSDS commented:
@jbalboni tests are all passing now!

jbalboni commented on src/applications/common/schemaform/save-in-progress/helpers.js:
```diff
@@ -0,0 +1,59 @@
+      path: 'form-saved',
+      component: FormSaved,
+      pageList,
+      config
```
Changing this from `formConfig` to `config` makes me a little nervous. The non-protected routes still expect and use `formConfig`. I think this might trip us up later.

annekainicUSDS replied:
no worries, I'll change it back! also noticing I have some commented out lines of code above, will remove that too

Commit:
Revert change from formConfig to config

annekainicUSDS commented:
Thanks so much for all of your reviews on this @jbalboni! I'm going to give everyone one more heads up tomorrow morning before I merge this.

cehsu reviewed (approved):
LGTM.
